beluga/include/beluga/sensor/data/sparse_value_grid.hpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_SENSOR_DATA_SPARSE_VALUE_GRID_HPP
+#define BELUGA_SENSOR_DATA_SPARSE_VALUE_GRID_HPP
+
+#include <Eigen/Core>
+#include <type_traits>
+
+#include <beluga/sensor/data/regular_grid.hpp>
+/**
+ * \file
+ * \brief Implementation of generic dense value grid.
+ */
+
+namespace beluga {
+
+/// Generic 2D sparse value grid.
+/**
+ * \tparam MapType:
+ *    Associative container representing a mapping from a Eigen::Vector2i to a value of type MapType::mapped_type.
+ *    It should implement a subset of standard library's associative containers public API.
+ *    In particular, given 'm' a possibly const instance of MapType:
+ *         - MapType::key_type must be Eigen::Vector2i.
+ *         - MapType::mapped_type should be the value type of the associative container entries.
+ *         - 'm.at(const Eigen::Vector2i& cell_index) const' should return a const reference to 'MapType::value_type'
+ * representing the value at that index, or throw 'std::out_of_range' if it doesn't exist.
+ *        - 'm.find(const Eigen::Vector2i&)' should follow the same API as
+ * [std::map](https://en.cppreference.com/w/cpp/container/map/find).
+ */
+template <typename MapType>
+class SparseValueGrid : public BaseRegularGrid2<SparseValueGrid<MapType>> {
```

Comment:
> Why is the current approach storage coupled? MapType can be anything that implements the requirements.

I see we are extracting cell and value types from storage.

> This would enforce MapType to be a templated class which I think it's an unnecessary restriction to add.

Fair point, though I see cell and value types changing much more often than the underlying storage, which will almost always be a hash map. Actually, have you considered [quadtrees](https://en.wikipedia.org/wiki/Quadtree) for this? They are much more efficient in terms of space, at the expense of ~ O(N log N) lookup complexity.

Reply:
As discussed in #309 , let's go for the easy O(1) lookup complexity for now.
 Moreover, the NDT representation already has lower memory footprint than its alternatives so it's not clear to me if we need to push even further in reducing that in detriment of lookup speed.